### DISCUSSION TITLE
Lesson 7: A confusion about `struct` in Solidity

### DISCUSSION THREAD

**Discussion by TreapGoGo** [Cyfrin/foundry-full-course-cu#178] [Q&A] [answered]
In `DeployFundMe.s.sol` we have

```solidity
address ethUsdPriceFeed = helperConfig.activeNetworkConfig();
```

However, in `HelperConfig.s.sol` we have

```solidity
contract HelperConfig is Script {
    NetworkConfig public activeNetworkConfig;
    struct NetworkConfig {
        address priceFeed;
    }
    ......
}
```

In my point of view, `helperConfig.activeNetworkConfig` is an instance of `NetworkConfig`. It is an variable, instead of a function or a method.

So why is there a pair of brackets at the end of `helperConfig.activeNetworkConfig()`?

-----------

I tried to write in other ways, but all of them got error.

Just like some other Objective-Oriented languages, I attempted to write

```solidity
address ethUsdPriceFeed = helperConfig.activeNetworkConfig.priceFeed;
```

The error is 

```
Member "priceFeed" not found or not visible after argument-dependent lookup in function () view external returns (address).solidity(9582)
```

----------------

ChatGPT told me that "Solidity interprets the brackets as a function call. The brackets () are used to indicate that you want to access a specific field within the returned struct." It suggests that I should write like this:

```solidity
address ethUsdPriceFeed = helperConfig.activeNetworkConfig().priceFeed;
```

However, the error is the same as above.

-------------------

In the video [50:30](https://youtu.be/sas02qSFZ74?t=3026), Patrick mistakenly code as the last one under the recommendation of Copilot. He just said an excuse-me but didn't explain why. 

I would appreciate it very much if someone can tell me why only the first code is right and how exactly I can access the element of a struct in Solidity.

**Comment by shikhar229169** [comment]
All the state variables that are public, for them solidity automatically makes their getter function with the same name as that of variable.
So when you are trying to use the variable of HelperConfig in DeployFundMe.s.sol, you can't access it via its name, you can access it via its getter function.

And to access the members for the activeNetworkConfig you have to write like this
```
(address priceFeeds) =  helperConfig.activeNetworkConfig();
```
Suppose the struct NetworkConfig struct looks like.
```
struct NetworkConfig {
        address priceFeed;
        uint256 value;
}
```

Then you can access it in DeployFundMe.s.sol by:
```
(address priceFeeds, uint256 val) =  helperConfig.activeNetworkConfig();
```

**Reply by TreapGoGo** [reply]
Thanks so much for your detailed explanation!

**Comment by pyromaus** [comment, marked as answer]
solidity automatically makes little getter functions for the variables you store so you can access them externally.
```solidity
address ethUsdPriceFeed = helperConfig.activeNetworkConfig().priceFeed;
```
trying to access the 'priceFeed' part of your NetworkConfig type variable doesn't work above cos the getter function for this has a simple end goal: to return the entire struct. It doesn't let you choose what to access before doing that.
If you change HelperConfig.sol by adding a second element to the config like so..
```solidity
struct NetworkConfig {
        address priceFeed; // ETHUSD price feed
        uint chainId; // chain ID
    }
.....
NetworkConfig memory anvilConfig = NetworkConfig(address(mockPriceFeed), block.chainid);
        return anvilConfig;
(address ethUsdPriceFeed, uint chainId) = helperConfig.activeNetworkConfig();
```
The getter will return the entire struct. Only after can you choose which elements to use.
If I wanna store only the chainId, scrap the address and leave the comma
```solidity
(, uint chainId) = helperConfig.activeNetworkConfig();
```
On the contrary, if you are calling the struct from inside the originating contract, you can access it that way without a problem:
```solidity
activeNetworkConfig.priceFeed
activeNetworkConfig.chainId
```
Cos while inside the contract we are accessing the actual struct and not just a pointer to it

**Reply by TreapGoGo** [reply]
Thanks so much for your detailed explanation!

**Reply by PatrickAlphaC** [reply]
nice work!

**Reply by wilson0922** [reply]
Nice discussion! When I reviewed this part, I got the same confusion. Thank you guys!!
